### PR TITLE
Use Haskell variables for passing values.

### PR DESCRIPTION
**Pull request by MarcFontaine**
This is a combination of several commits. (They were squashed into one big commit during an elaborate re-base).
Most changes are related to the way filenames work in `cardano-testnet`.

Currently the `cardano-testnet` uses filenames (Strings) to reference objects like keys, certs etc.
Two locations in the code refer to the same object by "magically" using the same filename.
This means that the data flow is implicit and hard to follow.
Old scheme:
```
do
cli_call_1 "filenameX" "filenameY"
cli_call_2 "filenameX" "filenameZ"
cli_call_3 "filenameA" "filenameB"
```
This PR changes the design as follows:
* File names are stored in an opaque `File` datatype (which wraps the file path).
* A wrapper is used for CLI calls. When a CLI call creates a file the wrapper returns the corresponding `File` resource.
* While CLI calls have type `IO ()` the wrapper has type `IO (File,..,File).
* CLI calls don't construct magic Filenames, but instead use the `File` resource that was created earlier.
The new scheme is
```
do
(file1, file2) <- cli_call1 arg1 arg2     -- cli_call1 creates file resources file1 and file2
(fileX, fileY) <- cli_call2 file1 file2    -- cli_call2 uses file resourcs file1 and file2
```
Full transformation to new scheme for filenames is still WIP.

### CODE REVIEW

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
There are a couple issues with this PR that need to be addressed before it is reviewed:
- There are several left over comments that should not be there
- There is a single commit for a relatively large diff

You need to remove comments that are no longer required and break this PR up in to several commits with clear commit messages describing the changes in the given commit. This makes it easier for a reviewer to understand the particular changes you have made.

File: cardano-testnet/src/Testnet/Cardano.hs [hunk 1]
```diff
@@ -1,7 +1,11 @@
 {-# LANGUAGE NamedFieldPuns #-}
 {-# LANGUAGE OverloadedStrings #-}
 {-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
 {-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-unused-do-bind #-}
+{-# OPTIONS_GHC -Wno-unused-matches #-}
+{-# OPTIONS_GHC -Wno-unused-top-binds #-}
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:8]: Why do we need `no-unused-matches` and `no-unused-top-binds`?

File: cardano-testnet/src/Testnet/Cardano.hs [hunk 2]
```diff
@@ -185,30 +191,104 @@ mkTopologyConfig numNodes allPorts port True = J.encode topologyP2P
...
+       testnetOptions
+       (configurationTemplate configuration)
+       startTime
+{-
+  -- dead code: not checked in any test
+  cardanoTestnetByronGenesisExpenditure testnetOptions configuration byronGenesisKeys
+  cardanoTestnetByronGovernance
+    (L.length $ cardanoBftNodeOptions testnetOptions)
+    testnetMagic
+    tmpDir
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:221]: Can you elaborate why you mean be "dead code".  There are a number of such references.

File: cardano-testnet/src/Testnet/Cardano.hs [hunk 3]
```diff
@@ -185,30 +191,104 @@ mkTopologyConfig numNodes allPorts port True = J.encode topologyP2P
...
+  -> CardanoTestnetOptions
+  -> FilePath
+  -> DTC.UTCTime
+  -> H.Integration
+    ( [File ByronKey]
+    , [File ByronDelegationCert]
+    , [File ByronDelegationKey]
+    )
+cardanoTestnetByronGenesis tp testnetMagic testnetOptions configurationTemplate startTime = do
+  let (TmpPath tmpDir) = tp
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:278]: Redundant parentheses.

File: cardano-testnet/src/Testnet/Cardano.hs [hunk 4]
```diff
@@ -268,20 +348,20 @@ cardanoTestnet testnetOptions H.Conf {..} = do
...
     let port = fromJust $ M.lookup node nodeToPort
-    H.lbsWriteFile (tempAbsPath </> node </> "topology.json") $
+    H.lbsWriteFile (tmpDir </> node </> "topology.json") $
       mkTopologyConfig (numBftNodes + cardanoNumPoolNodes (cardanoNodes testnetOptions))
                        allPorts port (cardanoEnableP2P testnetOptions)
 
-    H.writeFile (tempAbsPath </> node </> "port") (show port)
 
-  H.lbsWriteFile (tempAbsPath </> "byron.genesis.spec.json")
+    H.writeFile (tmpDir </> node </> "port") (show port)
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:363]: I think adding a newline here adds clarity.

File: cardano-testnet/test/Test/FoldBlocks.hs [hunk 5]
```diff
@@ -35,13 +35,12 @@ instance Show FoldBlocksException where
...
 prop_foldBlocks = H.integrationRetryWorkspace 2 "foldblocks" $ \tempAbsBasePath' -> do
-
   -- Start testnet
   base <- HE.noteM $ liftIO . IO.canonicalizePath =<< HE.getProjectBase
-  configTemplate <- H.noteShow $ base </> "configuration/defaults/byron-mainnet/configuration.yaml"
+  configTemplate <- H.noteShow $ base </> "configuration/defaults/byron-mainnet/configuration.yaml" --- outside of the package directory !!
   conf <- HE.noteShowM $
     TN.mkConf (TN.ProjectBase base) (TN.YamlFilePath configTemplate)
-      (tempAbsBasePath' <> "/")
+      (tempAbsBasePath' <> "/") --  '<> "/")'  is needed otherwise test fails !
```
Comment (newhoggy) [inline cardano-testnet/test/Test/FoldBlocks.hs:43]: I know this is old code, but can you find out why?

File: cardano-testnet/test/Test/FoldBlocks.hs [hunk 6]
```diff
@@ -35,13 +35,12 @@ instance Show FoldBlocksException where
 -- that main thread blocks on.
 prop_foldBlocks :: H.Property
 prop_foldBlocks = H.integrationRetryWorkspace 2 "foldblocks" $ \tempAbsBasePath' -> do
-
   -- Start testnet
   base <- HE.noteM $ liftIO . IO.canonicalizePath =<< HE.getProjectBase
-  configTemplate <- H.noteShow $ base </> "configuration/defaults/byron-mainnet/configuration.yaml"
+  configTemplate <- H.noteShow $ base </> "configuration/defaults/byron-mainnet/configuration.yaml" --- outside of the package directory !!
```
Comment (newhoggy) [inline cardano-testnet/test/Test/FoldBlocks.hs:40]: > outside of the package directory !!

Yeah, that's what `base` and `getProjectBase` means.  Is there a way to make this clearer without adding a comment?

File: cardano-testnet/src/Testnet/Util/Runtime.hs [hunk 7]
```diff
@@ -117,22 +122,38 @@ readNodeLoggingFormat = \case
...
+  { unTmpPath :: FilePath
+  } deriving (Eq, Show)
+getTmpRelPath :: TmpPath -> FilePath
+getTmpRelPath (TmpPath fp) = FP.makeRelative (getTmpBaseAbsPath (TmpPath fp)) fp
+getSocketDir :: TmpPath -> FilePath
+getSocketDir fp = getTmpRelPath fp </> "socket"
+getTmpBaseAbsPath :: TmpPath -> FilePath
+getTmpBaseAbsPath (TmpPath fp) = FP.takeDirectory fp
+getLogDir :: TmpPath -> FilePath
+getLogDir (TmpPath fp) = fp </> "logs"
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Runtime.hs:141]: Looks a bit crowded.  Can you spread them out with newlines?

File: cardano-testnet/src/Testnet/Util/Runtime.hs [hunk 8]
```diff
@@ -117,22 +122,38 @@ readNodeLoggingFormat = \case
 allNodes :: TestnetRuntime -> [NodeRuntime]
 allNodes tr = bftNodes tr <> fmap poolRuntime (poolNodes tr)
 
+makeSprocket
+  :: TmpPath
+  -> String
+  -> Sprocket
+makeSprocket tmpPath node
+  = Sprocket (getTmpBaseAbsPath tmpPath)  (getSocketDir tmpPath </> node)
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Runtime.hs:130]: Double space between `)  (` could be one line.

File: cardano-testnet/src/Testnet/Util/Runtime.hs [hunk 9]
```diff
@@ -67,16 +72,16 @@ data NodeRuntime = NodeRuntime
 
 data PoolNode = PoolNode
   { poolRuntime :: NodeRuntime
-  , poolKeys :: PoolNodeKeys
+  , poolKeys :: PoolNodeKeys -- used in Alonzo LeadershipShedule
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Runtime.hs:75]: I'm curious about the comment.  We shouldn't have to care where this data is used.  For example new tests might be added that make use of this.

File: cardano-testnet/src/Testnet/Util/Runtime.hs [hunk 10]
```diff
@@ -67,16 +72,16 @@ data NodeRuntime = NodeRuntime
...
   }
 
 data PoolNodeKeys = PoolNodeKeys
-  { poolNodeKeysColdVkey :: FilePath
+  { poolNodeKeysColdVkey :: FilePath -- == Operator keys -- used in LeadershipShedule
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Runtime.hs:79]: I'm partial to using something more type safe than `FilePath` for each type of file to make usage information easier to find but it doesn't have to be this PR.

File: cardano-testnet/src/Testnet/Cardano.hs [hunk 11]
```diff
@@ -298,98 +378,134 @@ cardanoTestnet testnetOptions H.Conf {..} = do
...
   -- Symlink the BFT operator keys from the genesis delegates, for uniformity
   forM_ bftNodesN $ \n -> do
-    H.createFileLink (tempAbsPath </> "byron/delegate-keys.00" <> show @Int (n - 1) <> ".key") (tempAbsPath </> "node-bft" <> show @Int n </> "byron/delegate.key")
-    H.createFileLink (tempAbsPath </> "byron/delegation-cert.00" <> show @Int (n - 1) <> ".json") (tempAbsPath </> "node-bft" <> show @Int n </> "byron/delegate.cert")
+    H.createFileLink (tmpDir </> "byron/delegate-keys.00" <> show @Int (n - 1) <> ".key") (tmpDir </> "node-bft" <> show @Int n </> "byron/delegate.key")
+    H.createFileLink (tmpDir </> "byron/delegation-cert.00" <> show @Int (n - 1) <> ".json") (tmpDir </> "node-bft" <> show @Int n </> "byron/delegate.cert")
+  let
+    forallBftNodesMkFile x = forM [1..numBftNodes] (fakeItH . x)
+
+  (genesisKeys :: [File ByronKey])<- forallBftNodesMkFile $ \n -> "byron/genesis-keys.00" <> show @Int (n - 1) <> ".json"
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:396]: Can the typing be done by type applications instead of type ascriptions?

File: cardano-testnet/src/Testnet/Cardano.hs [hunk 12]
```diff
@@ -298,98 +378,134 @@ cardanoTestnet testnetOptions H.Conf {..} = do
...
+    , "--tx", tempAbsPath </> "tx0.tx"
+    , "--wallet-key", tempAbsPath </> "byron/delegate-keys.000.key"
+    , "--rich-addr-from", richAddrFrom
+    , "--txout", show @(String, Int) (txAddr, fundsPerByronAddress)
+    ]
 
-  -- Update Proposal and votes
+{-
+cardanoTestnetByronGovernance creates two update-proposals and two update-votes.
+However they are never transmitted to the chain ?!
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Cardano.hs:447]: These were translated directly from the `mkfiles.sh` script this was based on.  Those scripts were designed so that the update proposals can be submitted manually.

We should consider what we want to do with this in the Haskell code.

File: cardano-testnet/src/Testnet/Util/Cli.hs [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+
+-- TODO: check no whitespace , slash etc allowed
+-- TODO: use Comment instead of Filepath
+newtype Comment = Comment { unComment :: String}
+  deriving (Show, Eq, Ord )
+
+instance IsString Comment where fromString a = Comment a
+
+instance Semigroup Comment where
+  (Comment a) <> (Comment b) = Comment $ a <> "-" <> b
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:61]: Redundant parentheses.

File: cardano-testnet/src/Testnet/Util/Cli.hs [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+import qualified Hedgehog.Extras.Test.File as H (writeFile)
+import           Hedgehog
+import           Hedgehog.Gen
+
+import           Testnet.Util.Process
+
+-- TODO: check no whitespace , slash etc allowed
+-- TODO: use Comment instead of Filepath
+newtype Comment = Comment { unComment :: String}
+  deriving (Show, Eq, Ord )
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:56]: This will need to be MacOS and Windows friendly.  MacOS filenames are case insensitive.  Windows filenames are case-insensitive and many characters have special meaning.

Perhaps we can call it `Kebab` because it's using kebab-casing and we lock it down do only support lower case alpha-numerics.

File: cardano-testnet/src/Testnet/Util/Cli.hs [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+data OperatorCounter
+data ByronKey
+data ByronAddress
+data ByronDelegationKey
+data ByronDelegationCert
+
+type TmpDir = FilePath
+type TestnetMagic = Int
+newtype File a = File {unFile :: FilePath}
+  deriving (Show, Eq)
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:138]: Nice!  Can you add a newline above the `newtype`?

File: cardano-testnet/src/Testnet/Util/Cli.hs [hunk 16]
```diff
@@ -0,0 +1,186 @@
...
+makeFilePath path comment suffix = do
+  almostUnique <- Comment <$> replicateM 4 hexit
+  return $ path </> unComment ( comment <> almostUnique) <.> suffix
+
+data KeyNames = KeyNames
+  { verificationKeyFile :: FilePath
+  , signingKeyFile :: FilePath
+  }
+
+type KeyGen a = H.Integration (File (a VKey), File (a SKey))
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:74]: The `KeyGen` type alias seems less helpful.

File: cardano-testnet/src/Testnet/Util/Cli.hs [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+-}
+
+fakeIt :: FilePath -> File a
+fakeIt = File
+
+getVKeyPath ::  (File (a VKey), File (a SKey)) -> FilePath
+getVKeyPath (File a, _ ) = a
+
+getSKeyPath ::  (File (a VKey), File (a SKey)) -> FilePath
+getSKeyPath (_, File a) = a
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:160]: I think I rather just use `fst` and `snd` and `unFile`.

File: cardano-testnet/src/Testnet/Util/Cli.hs [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+-- When all uses of fakeItH are removed the cleanup is done.
+fakeItH :: FilePath -> H.Integration (File a)
+fakeItH filePath = do
+  return $ File filePath
+  {-
+  exists <- liftIO $ fileExists filePath
+  if exists
+    then return $ File filePath
+    else error (filePath <> " does not exist.")
+-}
```
Comment (newhoggy) [inline cardano-testnet/src/Testnet/Util/Cli.hs:151]: Commented out code.

**Comment by newhoggy** [comment]
I agree with @Jimbo4350 regarding breaking up the code into multiple commits.

File: cardano-testnet/src/Testnet/Cardano.hs (hunk 1 above)
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:8]: The refactor uses Haskell variables for file resources.
The CLI wrapper functions return those resources.
However the refactor is still WIP some resources are defined but not yet used.
Once everything is don't the warnings should be turned on again.

File: cardano-testnet/src/Testnet/Cardano.hs (hunk 2 above)
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:221]: Some parts of the code code call `cardano-cli` for example to create a transaction.
However that transaction is never ever transmitted to the testnet or reused.
Also the code is not tested, all the tests pass when the code is disabled.
`cardanoTestnetByronGovernance` creates genesis key, delegation certs and delegation keys,
however this are are not used.

File: cardano-testnet/test/Test/FoldBlocks.hs (hunk 5 above)
Comment (MarcFontaine) [reply cardano-testnet/test/Test/FoldBlocks.hs:43]: I only stumbled upon this because `(tempAbsBasePath' <> "/")` looks strange.
but as the comment says the tests fail without the trailing `<> "/"`.
I'm in the course of reordering the various file paths and this will probably be removed anyway.

File: cardano-testnet/test/Test/FoldBlocks.hs (hunk 6 above)
Comment (MarcFontaine) [reply cardano-testnet/test/Test/FoldBlocks.hs:40]: That comment is a TODO for me.
I think it is better to move copies of those file to the package itself
or maybe there is yet an other way to do this.

File: cardano-testnet/src/Testnet/Util/Runtime.hs (hunk 7 above)
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Runtime.hs:141]: ups yes some blank lines got messed up during the rebase

File: cardano-testnet/src/Testnet/Util/Runtime.hs (hunk 9 above)
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Runtime.hs:75]: I'm about to change the code so I naturally need to know where something is defined and where it is used.
I think I also found stuff that is defined but not not used.
IMHO most of the time, it is not a good idea to add record fields for data that is not used.
Just because the data might be used in the future is not a good reason.

File: cardano-testnet/src/Testnet/Cardano.hs (hunk 11 above)
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:396]: Thanks for spotting, those types can be inferred and annotation are superfluous.

File: cardano-testnet/src/Testnet/Cardano.hs [hunk 19]
```diff
@@ -298,98 +378,134 @@ cardanoTestnet testnetOptions H.Conf {..} = do
...
-      ] >>= H.writeFile (tempAbsPath </> "byron/genesis-address-00" <> show @Int (n - 1))
-
-  do
-    richAddrFrom <- S.firstLine <$> H.readFile (tempAbsPath </> "byron/genesis-address-000")
-    txAddr <- S.firstLine <$> H.readFile (tempAbsPath </> "byron/address-000")
+  byronAddress0:_ <- forM [1 .. numBftNodes] $ \n -> do
+    skey <- cliKeyGen tmpDir $ "byron/payment-keys.00" <> show @Int (n - 1) <> ".key"
+    cliSigningKeyAddress tmpDir testnetMagic skey $ "byron/address-00" <> show @Int (n - 1)
+
+  byronGensisAddress0:_ <- forM (zip [1 .. numBftNodes] genesisKeys) $ \(n,key) -> do
```
[the remark this answers is not in the document]
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:421]: There is more to be done here..its dead code inside dead code.
`forM [1 .. numBftNodes]` keys are generated but only the first is used.

File: cardano-testnet/src/Testnet/Cardano.hs (hunk 12 above)
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Cardano.hs:447]: Good information.
I should have a look at `mkfiles.sh`. Is it still available ? Is it still in use ?
I'll change that comment to a TODO:
There should be a test that submits the update proposals and checks that they are activated.
It was not clear to me, if that test had been removed or disabled, or if it had never existed.

File: cardano-testnet/src/Testnet/Util/Cli.hs (hunk 17 above)
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Cli.hs:160]: Interesting. I guess I had written the type signature `fst` copied it to a `snd` line
and then edited that line to become the function definition...

File: cardano-testnet/src/Testnet/Util/Cli.hs (hunk 18 above)
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Cli.hs:151]: Is there a simple function in the test frame-work that asserts that a file exists ?
Last time I looked at it there was only one which checks for a list of files.

**Comment by MarcFontaine** [comment]
> I agree with @Jimbo4350 regarding breaking up the code into multiple commits.

I break up my code into multiple commits anyway.
However, here it was impossible to rebase it as such, so I squashed the PR into one commit.
Overall this PR is very simple and not extremely long.
Did splitting up the PR again it make easier to review for you ?

File: cardano-testnet/src/Testnet/Util/Cli.hs (hunk 16 above)
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Util/Cli.hs:74]: I do not know.. I looked up the code `KeyGen` is used five times.

File: cardano-testnet/src/Testnet/Conf.hs [hunk 20]
```diff
@@ -2,43 +2,33 @@
...
   } deriving (Eq, Show)
 
-newtype YamlFilePath = YamlFilePath
+newtype YamlFilePath = YamlFilePath   -- from the project that surounds the cabal package
   { projectBase :: FilePath
   } deriving (Eq, Show)
 
 data Conf = Conf
   { tempAbsPath :: FilePath
-  , tempRelPath :: FilePath
```
Comment (Jimbo4350) [inline cardano-testnet/src/Testnet/Conf.hs:27]: I can 100% get behind this simplification :+1:

File: cardano-testnet/src/Testnet/Babbage.hs [hunk 21]
```diff
@@ -124,10 +125,12 @@ babbageTestnet testnetOptions H.Conf {..} = do
...
       { poolNodeKeysColdVkey = tempAbsPath </> "pools" </> "cold" <> show n <> ".vkey"
       , poolNodeKeysColdSkey = tempAbsPath </> "pools" </> "cold" <> show n <> ".skey"
+      , poolNodeKeysVrf = ( File $ tempAbsPath </> "node-spo" <> show n </> "vrf.vkey"
+                          , File $ tempAbsPath </> "node-spo" <> show n </> "vrf.skey"
+                          )
       , poolNodeKeysVrfVkey = tempAbsPath </> "node-spo" <> show n </> "vrf.vkey"
       , poolNodeKeysVrfSkey = tempAbsPath </> "node-spo" <> show n </> "vrf.skey"
-      , poolNodeKeysStakingVkey = tempAbsPath </> "pools" </> "staking-reward" <> show n <> ".vkey"
-      , poolNodeKeysStakingSkey = tempAbsPath </> "pools" </> "staking-reward" <> show n <> ".skey"
+      , poolNodeKeysOperator = undefined
```
[the remark this answers is not in the document]
Comment (MarcFontaine) [reply cardano-testnet/src/Testnet/Babbage.hs:133]: Actually, it got called in one of the tests and
the error message contained the exact Module and line number.
The problem is in the `TestnetRuntime` type, which can either fit `babbage` or `cardano` but not both eras.
(That whole block of code is to be removed in a followup PR)

**Comment by Jimbo4350** [comment]
> @Jimbo4350 @newhoggy Are there any comments or reviews on the underlying idea of the PR ? I see tons of reviews on `hlint` problems and unneeded parenthesis that could easily be fixed in a followup PR.

@MarcFontaine one the node-cli team we do not intentionally merge PRs with hlint problems and other style issues into `master`. You need to fix these first. I am going to separate the `TmpPath` change myself to demonstrate the types of PRs that I'm looking for.

**Comment by MarcFontaine** [comment]
> > @Jimbo4350 @newhoggy Are there any comments or reviews on the underlying idea of the PR ? I see tons of reviews on `hlint` problems and unneeded parenthesis that could easily be fixed in a followup PR.
> 
> @MarcFontaine on the node-cli team we do not intentionally merge PRs with hlint problems and other style issues into `master`. You need to fix these first. I am going to separate the `TmpPath` change myself to demonstrate the type of PRs that I'm looking for.

The job of hlint is best done by CI.
How about this:
1) If there are no technical issues with a PR it gets a review with an approval
2) If it also passes CI (including of cause the linter) it gets merged.
3) We happily discuss "other style issues" as long as it is fun and does not kill productivity 
4) "other style issues" don"t hold back a PR.

**Comment by Jimbo4350** [comment]
> > > @Jimbo4350 @newhoggy Are there any comments or reviews on the underlying idea of the PR ? I see tons of reviews on `hlint` problems and unneeded parenthesis that could easily be fixed in a followup PR.
> > 
> > 
> > @MarcFontaine on the node-cli team we do not intentionally merge PRs with hlint problems and other style issues into `master`. You need to fix these first. I am going to separate the `TmpPath` change myself to demonstrate the type of PRs that I'm looking for.
> 
> The job of hlint is best done by CI. How about this:
> 
>     1. If there are no technical issues with a PR it gets a review with an approval
> 
>     2. If it also passes CI (including of cause the linter) it gets merged.
> 
>     3. We happily discuss "other style issues" as long as it is fun and does not kill productivity
> 
>     4. "other style issues" don"t hold back a PR.

No. This PR can still be broken up and PRs need to be of a certain standard before merging. What you are suggesting will create more churn and make the code less readable. As I said I will demonstrate what I am looking for.

**Comment by MarcFontaine** [comment]
> No. This PR can still be broken up and PRs need to be of a certain standard, both technical and stylistic, before merging. What you are suggesting will create more churn and make the code less readable. As I said I will demonstrate what I am looking for.

The orginal code has been written two years ago and nobody would care about improving it or moving it forward.
I open a PR and try to propose some changes and suddenly people start to merge PR after PR that change right that code. Honestly, I do not care to much about demonstrating anything to anybody..
